src/foam/java/PropertyInfo.js

```diff
@@ -141,7 +141,14 @@ foam.CLASS({
             type: 'void',
             visibility: 'public',
             args: [{ name: 'o', type: 'Object' }, { name: 'value', type: 'Object' }],
-            body: '((' + this.sourceCls.name + ') o).' + this.setterName + '(cast(value));'
+            body: `
+              ${this.sourceCls.name} obj = (${this.sourceCls.name}) o;
+              obj.assert${foam.String.capitalize(this.propName)}(cast(value));
+              Object oldValue = obj.${this.propName}IsSet_ ? obj.${this.propName}_ : null;
+              obj.${this.propName}_ = obj.${this.propName}PreSet_(oldValue, cast(value));
```

Comment:
This approach is expensive because it involves created two new objects for every primitive set. In XHome we just inserted the code-snippet from the preset and postSet in the generated call rather than calling out to the method stored in the PropertyInfo. This avoided casting to objects and avoided any call overhead when pre/postset weren't defiined. Same for the assert above, so this code is now actually creating 3 objects.